### PR TITLE
added model definition conversion for llama3

### PR DESCRIPTION
**Pull request by wesleytruong**
## This pr adds a model state dict conversion between TT and HF.

It includes to and from huggingface, and importantly performs a permutation on the q and k attention matrices to address the differences in RoPE implementation between native LLaMA and HuggingFace. Thanks to @rlrs and @vwxyzjn for finding and helping to fix this issue #335, https://github.com/pytorch/torchtitan/issues/1291#issuecomment-2997077080

### Testing
I tested the correctness of the model conversion by using the two methods greedy decoding, and kl_divergence for thorough comparison.

To test the from_hf script I downloaded a model from HF hub, converted it using the script, and ran forward passes using torchtitan. To test the to_hf script I obtained original llama weights and used the verified llama->dcp script. Then I used the convert_to_hf script to convert these weights to safetensors checkpoint.

For kl_divergence I tested each to_hf and from_hf against the baseline hf model, and compared these to the to_hf and from_hf weights when not performing the permutation.
| permuted wq and wk | kl_div (hf->tt) | kl_div (tt->hf) |
| --- | --- | --- |
| ✅   | -3.8356e-15 | -1.4431e-14 |
| ❌   | 3.0658e-06 | 9.6463e-06 |

When comparing, we can clearly see the kl div loss is many orders of magnitude higher when not permuted, showing that these probability distributions don't accurately represent the baseline hf's probability distribution. However, due to the small amount of weights that need to be permuted in this case, the loss is still not very high in the incorrect case, and can be deceiving if only using this as the evaluation metric. Therefore we also use greedy decoding with long generated sequences, calculating the loss as the exact match ratio of generated tokens.
| permuted wq and wk | kl_div (hf->tt) | kl_div (tt->hf) |
| --- | --- | --- |
| ❌  |  |  |
| ✅  |  |  |

### Usage
The model conversion can be done in two ways. The first direct way is to use the new convert_from_hf.py or convert_to_hf.py script, but requires loading the entire model weights into cpu memory. The second way is to use the training config options to load/save in hf format during training.

This should bring us one step closer to https://github.com/pytorch/torchtitan/issues/1210

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -137,15 +137,15 @@ def __init__(self, job_config: JobConfig):
         )
 
         # build model (using meta init)
-        model_args = self.train_spec.model_args[job_config.model.flavor]
+        self.model_args = self.train_spec.model_args[job_config.model.flavor]
```
Comment (tianyu-l) [inline torchtitan/train.py]: This is a hacky way to let `CheckpointManager` to depend on `model_args`, indirectly via `train_states`.

Instead, let's make `StateDictAdapter` consumes the `model_args` during init, and change the static methods to instance methods.

File: torchtitan/models/llama3/model/state_dict_adapter.py [hunk 2]
```diff
@@ -4,18 +4,129 @@
...
+        dim = model_args.dim
+        head_dim = dim // n_heads
+        state_dict = {}
+
+        for key, value in hf_state_dict.items():
+            if "layers" in key:
+                abstract_key = re.sub(r"(\d+)", "{}", key, count=1)
+                layer_num = re.search(r"\d+", key).group(0)
+                new_key = Llama3StateDictAdapter.from_hf_map[abstract_key]
+                print(f"{new_key} in layer {layer_num}")
```
Comment (tianyu-l) [inline torchtitan/models/llama3/model/state_dict_adapter.py]: remove `print`

File: torchtitan/models/llama3/model/state_dict_adapter.py [hunk 3]
```diff
@@ -4,18 +4,129 @@
...
+        "model.layers.{}.self_attn.rotary_emb.inv_freq": None,
+        "model.layers.{}.mlp.gate_proj.weight": "layers.{}.feed_forward.w1.weight",
+        "model.layers.{}.mlp.up_proj.weight": "layers.{}.feed_forward.w3.weight",
+        "model.layers.{}.mlp.down_proj.weight": "layers.{}.feed_forward.w2.weight",
+        "model.layers.{}.input_layernorm.weight": "layers.{}.attention_norm.weight",
+        "model.layers.{}.post_attention_layernorm.weight": "layers.{}.ffn_norm.weight",
+        "model.norm.weight": "norm.weight",
+        "lm_head.weight": "output.weight",
+    }
+    to_hf_map = {v: k for k, v in from_hf_map.items()}
```
Comment (tianyu-l) [inline torchtitan/models/llama3/model/state_dict_adapter.py]: what happens to `"model.layers.{}.self_attn.rotary_emb.inv_freq": None,`?

File: torchtitan/models/llama3/model/state_dict_adapter.py [hunk 4]
```diff
@@ -4,18 +4,129 @@
...
+import re
 from typing import Any
 
 from torchtitan.protocols.state_dict_adapter import StateDictAdapter
 
+from .args import TransformerModelArgs
+
 
 class Llama3StateDictAdapter(StateDictAdapter):
+    from_hf_map = {
```
Comment (tianyu-l) [inline torchtitan/models/llama3/model/state_dict_adapter.py]: We can set this as a constant in this file, similar to https://github.com/pytorch/torchtune/blob/main/torchtune/models/llama4/_convert_weights.py#L63
instead of a class variable.

File: torchtitan/models/llama3/model/state_dict_adapter.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/llama3/model/state_dict_adapter.py]: we can do this in the `to_hf` method, since it's small, similar to https://github.com/pytorch/torchtune/blob/main/torchtune/models/llama4/_convert_weights.py#L223

**Review by tianyu-l — COMMENTED** [review]
great work! left some initial comments

File: scripts/convert_from_hf.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    model_args = train_spec.model_args[model_flavor]
+
+    config_manager = ConfigManager()
+    config = config_manager.parse_args(
+        [
+            "--model.tokenizer-path",
+            "./assets/tokenizer/Llama-3.1-8B",
+        ]
+    )
+    tokenizer = build_hf_tokenizer(config)
```
Comment (tianyu-l) [inline scripts/convert_from_hf.py]: you no longer need to do this after rebasing

File: torchtitan/models/llama3/model/state_dict_adapter.py (hunk 3 above)
Comment (wesleytruong) [reply torchtitan/models/llama3/model/state_dict_adapter.py]: > what happens to `"model.layers.{}.self_attn.rotary_emb.inv_freq": None,`?

This none mapping is just for reference but effectively does nothing. The RoPE weights in torchtitan will get dropped when mapping to huggingface due to the RoPE implementation differences

File: docs/checkpoint.md [hunk 6]
```diff
@@ -1,19 +1,9 @@
-## How to convert a Llama 3 checkpoint for use in torchtitan
+# How to use checkpoints in TorchTitan
```
Comment (tianyu-l) [inline docs/checkpoint.md]: let's use \`torchtitan\` (aka `torchtitan`) within the repo

File: docs/checkpoint.md [hunk 7]
```diff
@@ -52,30 +50,63 @@ last_save_model_only = true
...
-```
-python -m torch.distributed.checkpoint.format_utils dcp_to_torch torchtitan/outputs/checkpoint/step-1000 checkpoint.pt
+### PyTorch Meta Llama
+
+If you want to continue training from an existing model checkpoint, the checkpoint must be in the DCP format expected by the checkpoint manager.
+An example script for converting the original Llama3 checkpoints into the expected DCP format can be found in `scripts/convert_llama_to_dcp.py`.
+
+The script expects a path to the original checkpoint files, and a path to an output directory:
+```bash
+python -m scripts.convert_from_llama <input_dir> <output_dir>
```
Comment (tianyu-l) [inline docs/checkpoint.md]: let's move this to the bottom of this section -- from now on the recommended way would be to/from HF weights

File: docs/checkpoint.md [hunk 8]
```diff
@@ -52,30 +50,63 @@ last_save_model_only = true
...
 
-6. CONVERT SHARDED CHECKPOINTS TO A SINGLE FILE\
-Finally, once you have obtained the last checkpoint, you can use the following command to convert the sharded checkpoints to a single .pt file that can be loaded into torchtune:
+## Conversion support
 
-```
-python -m torch.distributed.checkpoint.format_utils dcp_to_torch torchtitan/outputs/checkpoint/step-1000 checkpoint.pt
+### PyTorch Meta Llama
+
+If you want to continue training from an existing model checkpoint, the checkpoint must be in the DCP format expected by the checkpoint manager.
```
Comment (tianyu-l) [inline docs/checkpoint.md]: this is not true anymore

File: docs/checkpoint.md [hunk 9]
```diff
@@ -52,30 +50,63 @@ last_save_model_only = true
...
+The script expects a path to the original checkpoint files, and a path to an output directory:
+```bash
+python -m scripts.convert_from_llama <input_dir> <output_dir>
 ```
 
-7. EXCLUDING SPECIFIC KEYS FROM CHECKPOINT LOADING
-In some cases, you may want to partially load from a previous-trained checkpoint and modify certain settings, such as the number of GPUs or the current step. To achieve this, you can use the `exclude_from_loading` parameter to specify which keys should be excluded from loading.
-This parameter takes a list of string that should be excluded from loading.
+
+### Torchtune
```
Comment (tianyu-l) [inline docs/checkpoint.md]: I think we can remove this section in general, as torchtune also supports HF checkpoints.

File: docs/checkpoint.md [hunk 10]
```diff
@@ -52,30 +50,63 @@ last_save_model_only = true
...
+interval = 10
+last_save_model_only = true
+export_dtype = "bfloat16"
 ```
-When used in command line, the parameter should be a comma-separated list of strings. For example: `--checkpoint.exclude_from_loading data_loader,lr_scheduler`.
+
+2. SAVE THE FINAL CHECKPOINT\
+Once the above have been set, the final checkpoint at the end of the training step will consist of model only with the desired export dtype. However, if the final step has not been reached yet, full checkpoints will still be saved so that training can be resumed.
+
+3. CONVERT SHARDED CHECKPOINTS TO A SINGLE FILE\
```
Comment (tianyu-l) [inline docs/checkpoint.md:99]: we can keep this part.
Basically in this section you'd have
1. HF conversion (conversion scripts or during training save/load)
2. the instruction here to convert DCP to torch (section title should be torch instead of torchtune)
3. alternative way of converting llama to DCP, which is to be deprecated

File: docs/checkpoint.md [hunk 11]
```diff
@@ -52,30 +50,63 @@ last_save_model_only = true
...
+2. To directly reformat the weights without the need to run a training loop, run the corresponding conversion script. The naming scheme is torchtitan-centric, e.g. convert_from_hf means convert hf->tt.
+
+```
+python ./scripts/checkpoint_conversion/convert_from_hf.py <input_dir> <output_dir> --model_name <model_name> --model_flavor <model_flavor>
+python ./scripts/checkpoint_conversion/convert_to_hf.py <input_dir> <output_dir> --model_name <model_name> --model_flavor <model_flavor>
+# e.g.
+python ./scripts/convert_from_hf.py ~/.cache/huggingface/hub/models--meta-llama--Meta-Llama-3-8B/snapshots/8cde5ca8380496c9a6cc7ef3a8b46a0372a1d920/ ./outputs/checkpoint/step-0 --model_name llama3 --model_flavor 8B
+```
+
+### Seed Checkpoint
```
Comment (tianyu-l) [inline docs/checkpoint.md]:
```suggestion
### Creating a seed checkpoint
```
This should be its own section, instead of being under conversion.

File: scripts/checkpoint_conversion/convert_from_hf.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    dcp.save(
+        state_dict,
+        checkpoint_id=output_dir,
+    )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert Llama weights to DCP format.")
+    parser.add_argument(
+        "input_dir", type=Path, help="Input directory with original Llama weights."
```
Comment (tianyu-l) [inline scripts/checkpoint_conversion/convert_from_hf.py]: HF checkpoints?

File: scripts/checkpoint_conversion/convert_to_hf.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+    config_path = output_dir / "config.json"
+    with config_path.open("w") as f:
+        json.dump(config_json, f, indent=4)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert Llama weights to HF format.")
+    parser.add_argument(
+        "input_dir", type=Path, help="Input directory with original Llama weights."
```
Comment (tianyu-l) [inline scripts/checkpoint_conversion/convert_to_hf.py]: DCP weights?

File: scripts/checkpoint_conversion/convert_to_hf.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        storage_writer=storage_writer,
+    )
+
+    config_path = output_dir / "config.json"
+    with config_path.open("w") as f:
+        json.dump(config_json, f, indent=4)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert Llama weights to HF format.")
```
Comment (tianyu-l) [inline scripts/checkpoint_conversion/convert_to_hf.py]: modify

File: scripts/checkpoint_conversion/convert_to_hf.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    parser.add_argument(
+        "input_dir", type=Path, help="Input directory with original Llama weights."
+    )
+    parser.add_argument("output_dir", type=Path, help="Output directory for DCP.")
```
Comment (tianyu-l) [inline scripts/checkpoint_conversion/convert_to_hf.py]: modify

File: torchtitan/train.py [hunk 16]
```diff
@@ -298,7 +298,11 @@ def __init__(self, job_config: JobConfig):
             lr_schedulers=self.lr_schedulers,
             states={"train_state": self},
             checkpoint_config=job_config.checkpoint,
-            sd_adapter=self.train_spec.state_dict_adapter,
+            sd_adapter=(
+                self.train_spec.state_dict_adapter(self.model_args)
```
Comment (tianyu-l) [inline torchtitan/train.py]: No need to make `model_args` a class variable
```suggestion
                self.train_spec.state_dict_adapter(model_args)
```
could you revert all the changes in this file, except for this section?

File: torchtitan/models/llama3/model/state_dict_adapter.py [hunk 17]
```diff
@@ -4,18 +4,145 @@
...
+        if self.model_args.ffn_dim_multiplier:
+            ffn_hidden_dim = int(ffn_hidden_dim * self.model_args.ffn_dim_multiplier)
+        multiple_of = self.model_args.multiple_of
+        ffn_hidden_dim = multiple_of * (
+            (ffn_hidden_dim + multiple_of - 1) // multiple_of
+        )  # hacky way to get ffn_hidden_dim, follows the calculation in models.TransformerBlock and model.FeedForward
+
+        config_json = {
+            "architectures": ["LlamaForCausalLM"],
+            "hidde": "silu",
```
Comment (tianyu-l) [inline torchtitan/models/llama3/model/state_dict_adapter.py]:
```suggestion
            "hidden_act": "silu",
```

File: torchtitan/models/llama3/model/state_dict_adapter.py [hunk 18]
```diff
@@ -4,18 +4,145 @@
...
+            "hidden_size": self.model_args.dim,
+            "intermediate_size": ffn_hidden_dim,
+            "model_type": "llama",
+            "num_attention_heads": self.model_args.n_heads,
+            "num_hidden_layers": self.model_args.n_layers,
+            "num_key_value_heads": self.model_args.n_kv_heads,
+            "vocab_size": self.model_args.vocab_size,
+        }
```
Comment (tianyu-l) [inline torchtitan/models/llama3/model/state_dict_adapter.py]: This seems only a partial set of https://huggingface.co/meta-llama/Meta-Llama-3-8B/blob/main/config.json
I think the value of such dict is little, compared with the complexity. I suggest we remove this in the output and only keep the `hf_state_dict`.

File: torchtitan/components/checkpoint.py [hunk 19]
```diff
@@ -402,6 +405,11 @@ def dcp_save(
                 checkpoint_id=checkpoint_save_id,
             )
 
+        if to_hf:
+            config_path = Path(checkpoint_id) / "config.json"
+            with config_path.open("w") as f:
+                json.dump(config_json, f, indent=4)
+
```
Comment (tianyu-l) [inline torchtitan/components/checkpoint.py]: In SPMD, multiple processes would run this I think. Not sure what will happen.
But anyways, let's not produce the `config.json` for the reason I mentioned in the other file.

File: docs/checkpoint.md (hunk 6 above)
Comment (tianyu-l) [inline docs/checkpoint.md]:
```suggestion
# How to use checkpointing in `torchtitan`
```

**Review by tianyu-l — COMMENTED** [review]
Have you tested correctness of save / load during training? It can be tricky because the value would be DTensors, instead of full plain tensors.
Also, could you test if save/load during training is correct when TP / PP is used?
PP is tricky because some module may not exists in certain PP ranks.

**Review by tianyu-l — APPROVED** [review]
LGTM! Great work!
